crates/bevy_pbr/src/render_graph/forward_pipeline/forward.frag

```diff
@@ -24,41 +27,170 @@ layout(set = 1, binding = 0) uniform Lights {
+    vec3 F = fresnel(f0, LoH);
+
+    return (D * V) * F;
+}
+
+float Fd_Burley(float roughness, float NoV, float NoL, float LoH) {
+    float f90 = 0.5 + 2.0 * roughness * LoH * LoH;
+    float lightScatter = F_Schlick(1.0, f90, NoL);
+    float viewScatter = F_Schlick(1.0, f90, NoV);
+    return lightScatter * viewScatter * (1.0 / PI);
+}
+
+vec3 computeDiffuseColor(const vec3 baseColor, float metallic) {
+    return baseColor * (1.0 - metallic);
+}
+
+#define MIN_N_DOT_V 1e-4
+
+float clampNoV(float NoV) {
+    // Neubelt and Pettineo 2013, "Crafting a Next-gen Material Pipeline for The
+    // Order: 1886"
+    return max(NoV, MIN_N_DOT_V);
+}
+
+float perceptualRoughnessToRoughness(float perceptualRoughness) {
+    return perceptualRoughness * perceptualRoughness;
+}
+
+#if defined(TARGET_MOBILE)
+// min roughness such that (MIN_PERCEPTUAL_ROUGHNESS^4) > 0 in fp16 (i.e.
+// 2^(-14/4), rounded up)
+#define MIN_PERCEPTUAL_ROUGHNESS 0.089
+#define MIN_ROUGHNESS 0.007921
+#else
+#define MIN_PERCEPTUAL_ROUGHNESS 0.045
+#define MIN_ROUGHNESS 0.002025
+#endif
+
+#endif
 
 void main() {
     vec4 output_color = Albedo;
-# ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
-    output_color *= texture(
-        sampler2D(StandardMaterial_albedo_texture, StandardMaterial_albedo_texture_sampler),
-        v_Uv);
-# endif
-
-# ifndef STANDARDMATERIAL_UNLIT
-    vec3 normal = normalize(v_Normal);
+#ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
+    output_color *= texture(sampler2D(StandardMaterial_albedo_texture,
+                                      StandardMaterial_albedo_texture_sampler),
+                            v_Uv);
+#endif
+
+#ifndef STANDARDMATERIAL_UNLIT
+    float perceptual_roughness = pbr.x;
+    perceptual_roughness =
+        clamp(perceptual_roughness, MIN_PERCEPTUAL_ROUGHNESS, 1.0);
+    float roughness = perceptualRoughnessToRoughness(perceptual_roughness);
+
+    float metallic = pbr.y;
+    vec3 N = normalize(v_Normal);
+    vec3 V = normalize(CameraPos.xyz - v_Position.xyz);
+
+    vec3 F0 = vec3(0.04);
+    F0 = mix(F0, output_color.rgb, metallic);
+
+    vec3 diffuseColor = computeDiffuseColor(output_color.rgb, metallic);
+
     // accumulate color
-    vec3 color = AmbientColor;
-    for (int i=0; i<int(NumLights.x) && i<MAX_LIGHTS; ++i) {
+    vec3 light_accum = vec3(0.0);
+    for (int i = 0; i < int(NumLights.x) && i < MAX_LIGHTS; ++i) {
         Light light = SceneLights[i];
-        // compute Lambertian diffuse term
-        vec3 light_dir = normalize(light.pos.xyz - v_Position);
-        float diffuse = max(0.0, dot(normal, light_dir));
-        // add light contribution
-        color += diffuse * light.color.xyz;
+
+        vec3 lightDir = light.pos.xyz - v_Position.xyz;
+        vec3 L = normalize(lightDir);
+
+        float rangeAttenuation =
+            getDistanceAttenuation(lightDir, light.attenuation);
+
+        vec3 H = normalize(L + V);
+
+        float NdotL = clampNoV(dot(N, L));
```

Comment:
You'll probably want to consolidate `NdotL` and `NoL` since they are essentially the same value with redundant steps.

Reply:
I don't think NoL needs to be clamped to `[1e-4, 1]` at all.